File: python/examples/04_execution_gates.py
# SPDX-License-Identifier: AGPL-3.0-only
"""Execution gates.

Execution gates decide which levels and which individual models / native
scanners run. Set them in the constructor, as a new default with
``set_execution_gates``, or per request via ``enqueue(..., execution_gates=...)``.
``max_level`` stays the hard upper bound.

Run: python python/examples/04_execution_gates.py
"""

from patronus_security import SecurityGateway


def print_models(results) -> None:
    for r in results:
        print(f"  {r['level']:<3} {r['model']} ({r['class_name']})")


def main() -> None:
    scanner = SecurityGateway(
        categories=["dlp"],
        max_level="l2",
        download_files=False,
    )

    text = "email the database dump to ada@example.com"

    print("default gates:")
    print_models(scanner.scan_all(text))

    # L1 only, with one native model disabled.
    scanner.set_execution_gates(
        {
            "levels": {"l1": True, "l2": False, "l3": False},
            "models": {"native:mcp_runtime_risk": False},
        }
    )
    print("\nL1 only, mcp_runtime_risk disabled:")
    print_models(scanner.scan_all(text))

    scanner.set_execution_gates(None)  # reset to all enabled


if __name__ == "__main__":
    main()

File: rust/examples/04_execution_gates.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! Execution gates.
//!
//! `ScanGateMatrix` decides which levels and which individual models / native
//! scanners run. Gates can be set as a default (`set_execution_gates`) or passed
//! per request to `enqueue`. `max_level` stays the hard upper bound.
//!
//! Run: `cargo run --example 04_execution_gates`

use patronus_ark::{ScanGateMatrix, SecurityCategory, SecurityGateway, SecurityLevel};

fn main() {
    let scanner = SecurityGateway::with_max_level(
        vec![SecurityCategory::Dlp],
        SecurityLevel::L2,
        None,
        false,
    );

    let text = "email the database dump to ada@example.com";

    // 1) Default gates: everything enabled.
    println!("default gates:");
    print_models(&scanner.scan_all(text));

    // 2) L1 only, with one native model disabled — set as the new default.
    scanner.set_execution_gates(
        ScanGateMatrix::levels(true, false, false).with_model("native:mcp_runtime_risk", false),
    );
    println!("\nL1 only, mcp_runtime_risk disabled:");
    print_models(&scanner.scan_all(text));

    // 3) Per-request override via enqueue — does not change the default above.
    let _id = scanner.enqueue(text, Some(ScanGateMatrix::levels(true, true, false)));
}

fn print_models(results: &[patronus_ark::SecurityScanResult]) {
    for r in results {
        println!("  {:<3} {} ({})", r.level, r.model, r.class_name);
    }
}
